core/src/main/java/io/grpc/transport/AbstractStream.java

```diff
@@ -229,6 +278,74 @@ protected final void requestMessagesFromDeframer(int numMessages) {
     }
   }
 
+  /**
+   * Event handler to be called by the subclass when a number of bytes are being queued for sending
+   * to the remote endpoint. May call back the listener's {@link StreamListener#onReady()}
+   * handler if appropriate.
+   *
+   * @param numBytes the number of bytes being sent.
+   */
+  protected final void onSendingBytes(int numBytes) {
+    boolean shouldNotify = false;
+    synchronized (onReadyLock) {
+      shouldNotifyOnReady = true;
+      numSentBytesQueued += numBytes;
+      shouldNotify = shouldNotifyOnReady();
+    }
+    if (shouldNotify) {
+      listener().onReady();
```

Comment:
My feeling is that would lead to a lot of recursion, such that it could exhaust the stack. At 32K, we can fit 6.5K messages in the buffer.

If we are doing it for efficiency, then I think that returning a boolean would be a better approach. Doing a callback requires multiple allocations, has the reentrancy/recursion problem, has more synchronization in grpc, and needs more application synchronization between sender and listener. From a threading perspective, only doing callbacks from transport thread would also be a plus.